plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -735,3 +735,188 @@ func haveSame(a1, a2 []string) bool {
+		podAntiAffinityTermMatches, err := checkAnyPodThatMatchPodAffinityTerm(pod, existingPods, node, podAntiAffinityTerm)
+		if err != nil || podAntiAffinityTermMatches == true {
+			glog.V(10).Infof("Cannot schedule pod %+v onto node %v, because not all the existing pods on this node satisfy the PodAntiAffinityTerm %v,err: %v",
+				podName(pod), node.Name, podAntiAffinityTerm, err)
+			return false
+		}
+	}
+
+	// Check if scheduling the pod onto this node would break
+	// any anti-affinity rules indicated by the existing pods on the node.
+	// If it would break, system should not schedule pod onto this node.
+	podsToBeOnNode := []*api.Pod{pod}
+	for _, ep := range existingPods {
+		epAffinity, err := api.GetAffinityFromPodAnnotations(ep.Annotations)
+		if err != nil {
+			glog.V(10).Infof("Failed to get Affinity from Pod %+v, err: %+v", podName(pod), err)
+			return false
+		}
+		if epAffinity.PodAntiAffinity != nil {
+			var epAntiAffinityTerms []api.PodAffinityTerm
+			if len(epAffinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution) != 0 {
+				epAntiAffinityTerms = epAffinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution
+			}
+			// TODO: Uncomment this block when implement RequiredDuringSchedulingRequiredDuringExecution.
+			//if len(epAffinity.PodAntiAffinity.RequiredDuringSchedulingRequiredDuringExecution) != 0 {
+			//	epAntiAffinityTerms = append(epAntiAffinityTerms, epAffinity.PodAntiAffinity.RequiredDuringSchedulingRequiredDuringExecution...)
+			//}
+
+			for _, epAntiAffinityTerm := range epAntiAffinityTerms {
+				existingPodAntiAffinityTermMatches, err := checkAnyPodThatMatchPodAffinityTerm(ep, podsToBeOnNode, node, epAntiAffinityTerm)
+				// if err or podAffinity match passes then return false otherwise continue
+				if err != nil || existingPodAntiAffinityTermMatches == true {
+					glog.V(10).Infof("Cannot schedule Pod %+v, onto node %v because the pod does not satisfy the PodAntiAffinityTerm%+v,of existing pod %+v,err: %v",
+						podName(pod), node.Name, epAntiAffinityTerm, podName(ep), err)
+					return false
+				}
+			}
+		}
+	}
+	// all the required pod anti-affinity scheduling rules are satisfied
+	glog.V(10).Infof("Can schedule Pod %+v, on node %v because all the required pod anti-affinity scheduling rules are satisfied", podName(pod), node.Name)
+	return true
+}
+
+// nodeMatchPodAffinityAntiAffinity checks if the node matches
+// the requiredDuringScheduling affinity/anti-affinity rules indicated by the pod.
+func nodeMatchPodAffinityAntiAffinity(pod *api.Pod, existingPods []*api.Pod, node *api.Node, allPods []*api.Pod) bool {
+	// Parse required affinity scheduling rules.
+	affinity, err := api.GetAffinityFromPodAnnotations(pod.Annotations)
+	if err != nil {
+		glog.V(10).Infof("Failed to get Affinity from Pod %+v, err: %+v", podName(pod), err)
+		return false
+	}
+
+	// check if the current node match the inter-pod affinity scheduling rules.
+	if affinity.PodAffinity != nil {
```

Comment:
The code from here to the end of the function is correct, but seems a little overly-verbose. One slightly more concise way to write it is

```
if affinity.PodAffinity != nil {
   if !checkNodeMatchesHardPodAffinity(pod, existingPods, node, allPods, affinity.PodAffinity) {
       return false
   }
}
if affinity.PodAntiAffinity != nil {
   if !checkNodeMatchesHardPodAntiAffinity(pod, existingPods, node, affinity.PodAntiAffinity) {
      return false
   }
}
return true
```